util/cpuprofile/cpuprofile.go

```diff
@@ -0,0 +1,219 @@
+	globalCPUProfiler.register(ch)
+}
+
+// Unregister unregister a ProfileConsumer from the global CPU profiler.
+// The unregistered ProfileConsumer won't receive the cpu profile data any more.
+// This function is thread-safe.
+func Unregister(ch ProfileConsumer) {
+	globalCPUProfiler.unregister(ch)
+}
+
+// parallelCPUProfiler is a cpu profiler.
+// With parallelCPUProfiler, it is possible to have multiple profile consumer at the same time.
+// WARN: Only one running parallelCPUProfiler is allowed in the process, otherwise some profiler may profiling fail.
+type parallelCPUProfiler struct {
+	sync.Mutex
+	cs             map[ProfileConsumer]struct{}
+	notifyRegister chan struct{}
+
+	// some data cache for profiling.
+	data         *ProfileData
+	lastDataSize int
+
+	started bool
+	ctx     context.Context
+	cancel  context.CancelFunc
+	wg      sync.WaitGroup
+}
+
+// newParallelCPUProfiler crate a new parallelCPUProfiler.
+func newParallelCPUProfiler() *parallelCPUProfiler {
+	return &parallelCPUProfiler{
+		cs:             make(map[ProfileConsumer]struct{}),
+		notifyRegister: make(chan struct{}),
+	}
+}
+
+var (
+	errProfilerAlreadyStarted = errors.New("parallelCPUProfiler is already started")
+)
+
+func (p *parallelCPUProfiler) start() error {
+	p.Lock()
+	defer p.Unlock()
+	if p.started {
+		return errProfilerAlreadyStarted
+	}
+
+	p.started = true
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+	p.wg.Add(1)
+	go util.WithRecovery(p.profilingLoop, nil)
+
+	logutil.BgLogger().Info("parallel cpu profiler started")
+	return nil
+}
+
+func (p *parallelCPUProfiler) stop() {
+	p.Lock()
+	defer p.Unlock()
+	if !p.started {
+		return
+	}
+	p.started = false
+	if p.cancel != nil {
+		p.cancel()
+	}
+	p.wg.Wait()
+	logutil.BgLogger().Info("parallel cpu profiler stopped")
+}
+
+func (p *parallelCPUProfiler) register(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	p.cs[ch] = struct{}{}
+	p.Unlock()
+
+	select {
+	case p.notifyRegister <- struct{}{}:
+	default:
+	}
+}
+
+func (p *parallelCPUProfiler) unregister(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	delete(p.cs, ch)
+	p.Unlock()
+}
+
+func (p *parallelCPUProfiler) profilingLoop() {
+	checkTicker := time.NewTicker(defProfileDuration)
+	defer func() {
+		checkTicker.Stop()
+		pprof.StopCPUProfile()
+		p.wg.Done()
+	}()
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-p.notifyRegister:
+			if !p.inProfilingStatus() {
+				p.profileCycle()
+			}
+		case <-checkTicker.C:
+			p.profileCycle()
+		}
+	}
+}
+
+func (p *parallelCPUProfiler) profileCycle() {
+	if p.inProfilingStatus() {
+		pprof.StopCPUProfile()
+		p.lastDataSize = p.data.Data.Len()
+		p.sendToConsumers()
+	}
+
+	// Only do cpu profiling when there are consumers.
+	if p.consumersCount() > 0 {
+		metrics.CPUProfileCounter.Inc()
+		p.data = &ProfileData{Data: bytes.NewBuffer(make([]byte, 0, p.lastDataSize)), Begin: time.Now()}
+		err := pprof.StartCPUProfile(p.data.Data)
```

Comment:
When error happens when calling `StartCPUProfile`, you should not call `StopCPUProfile`, as it is a violation of the pprof API design.

Reply:
resolved, done.